File: rust-code-examples/src/design_patterns/factory.ts
// factory pattern

class Burger {
    private name: string;
    private price: number;
    private calories: number;

    constructor() {
        this.name = "";
        this.price = 0;
        this.calories = 0;
    }

    private createBurger(name: string, price: number, calories: number) {
        const burger = new Burger();
        burger.name = name;
        burger.price = price;
        burger.calories = calories;
        return burger;
    }

    createCheeseBurger() {
        return this.createBurger("Cheese Burger", 10, 200);
    }

    createChickenBurger() {
        return this.createBurger("Chicken Burger", 15, 250);
    }

    createFishBurger() {
        return this.createBurger("Fish Burger", 12, 220);
    }
}

const burgerFactory = new Burger();
const cheeseBurger = burgerFactory.createCheeseBurger();
const chickenBurger = burgerFactory.createChickenBurger();
const fishBurger = burgerFactory.createFishBurger();

File: rust-code-examples/src/design_patterns/factory.rs
#[non_exhaustive]
#[derive(Debug)]
struct Burger {
    patties: i32,
    bacon: bool,
    cheese: bool,
    sauce: String,
    toasted: bool,
}

impl Burger {
    fn create_burger(
        patties: i32,
        bacon: bool,
        cheese: bool,
        sauce: &str,
        toasted: bool,
    ) -> Burger {
        Burger {
            patties,
            bacon,
            cheese,
            sauce: sauce.to_string(),
            toasted,
        }
    }

    fn create_cheeseburger(patties: i32, bacon: bool, toasted: bool) -> Burger {
        Burger::create_burger(patties, bacon, true, "tomato", toasted)
    }

    fn create_baconburger(patties: i32, toasted: bool) -> Burger {
        Burger::create_burger(patties, true, false, "tomato", toasted)
    }

    fn create_veggieburger(patties: i32, toasted: bool) -> Burger {
        Burger::create_burger(patties, false, false, "tomato", toasted)
    }
}

fn factory() {
    let cheeseburger = Burger::create_cheeseburger(2, true, true);
    let baconburger = Burger::create_baconburger(3, true);
    let veggieburger = Burger::create_veggieburger(1, false);

    println!("cheeseburger: {:?}", cheeseburger);
    println!("baconburger: {:?}", baconburger);
    println!("veggieburger: {:?}", veggieburger);
}
